File: frontend/src/utils/foodEstimator.js
import { getApiBaseUrl } from "../constants/config";
import { formatPortionNote, parsePortion, scaleNutrition, stripPortionSuffix } from "./portionParser";
import { parseFoodTextLocally } from "./nlpFoodParser";

const DEFAULT_NUTRITION = { calories: 450, protein: 20, carbs: 52, fat: 16 };
const THAI_FOODS_URL = `${process.env.PUBLIC_URL || ""}/data/thai_foods.json`;

let thaiFoodsCache = null;
let aliasIndexCache = null;

const normalize = (text) => String(text || "").trim().toLowerCase().replace(/\s+/g, " ");

const similarity = (a, b) => {
  if (!a || !b) return 0;
  if (a === b) return 1;
  if (a.includes(b) || b.includes(a)) {
    return 0.82 + (Math.min(a.length, b.length) / Math.max(a.length, b.length)) * 0.18;
  }

  const matrix = Array.from({ length: a.length + 1 }, () => Array(b.length + 1).fill(0));
  for (let i = 0; i <= a.length; i += 1) matrix[i][0] = i;
  for (let j = 0; j <= b.length; j += 1) matrix[0][j] = j;

  for (let i = 1; i <= a.length; i += 1) {
    for (let j = 1; j <= b.length; j += 1) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      matrix[i][j] = Math.min(
        matrix[i - 1][j] + 1,
        matrix[i][j - 1] + 1,
        matrix[i - 1][j - 1] + cost
      );
    }
  }

  const distance = matrix[a.length][b.length];
  return 1 - distance / Math.max(a.length, b.length);
};

export const loadThaiFoods = async () => {
  if (thaiFoodsCache) return thaiFoodsCache;
  try {
    const res = await fetch(THAI_FOODS_URL);
    if (!res.ok) return [];
    thaiFoodsCache = await res.json();
    return thaiFoodsCache;
  } catch {
    return [];
  }
};

const buildAliasIndex = (foods) => {
  if (aliasIndexCache) return aliasIndexCache;
  aliasIndexCache = foods
    .flatMap((item) => item.names.map((alias) => ({ alias, aliasNorm: normalize(alias), item })))
    .sort((a, b) => b.alias.length - a.alias.length);
  return aliasIndexCache;
};

const thaiFoodToSearchResult = (item, matchedAlias) => ({
  name: matchedAlias,
  baseName: item.names[0],
  calories: item.calories,
  protein: item.protein,
  carbs: item.carbs,
  fat: item.fat,
  defaultServingGrams: item.defaultServingGrams,
  defaultServingLabel: item.defaultServingLabel,
  per100g: item.per100g,
  category: item.category,
});

export const searchThaiFoods = async (query, limit = 12) => {
  const norm = normalize(query);
  if (!norm) return [];

  const foods = await loadThaiFoods();
  const scored = [];
  const seen = new Set();

  foods.forEach((item) => {
    const primary = item.names[0];
    if (seen.has(primary)) return;

    let bestAlias = null;
    let bestScore = 0;

    item.names.forEach((alias) => {
      const aliasNorm = normalize(alias);
      if (aliasNorm === norm) {
        bestAlias = alias;
        bestScore = 1;
        return;
      }
      const score = similarity(norm, aliasNorm);
      if (score > bestScore) {
        bestScore = score;
        bestAlias = alias;
      }
    });

    if (bestAlias && bestScore >= 0.55) {
      seen.add(primary);
      scored.push({ score: bestScore, result: thaiFoodToSearchResult(item, bestAlias) });
    }
  });

  return scored
    .sort((a, b) => b.score - a.score)
    .slice(0, limit)
    .map((row) => ({ ...row.result, _searchScore: row.score }));
};

const GENERIC_RICE_ALIASES = new Set(["ข้าวเปล่า", "ข้าวสวย"]);
const RICE_DISH_MARKERS = ["ยำ", "ผัด", "ราด", "มัน", "ซอย", "ต้ม", "หมก", "คลุก"];

const isGenericRiceMismatch = (queryNorm, matchedAlias, score) => {
  if (!GENERIC_RICE_ALIASES.has(matchedAlias)) return false;
  if (!RICE_DISH_MARKERS.some((marker) => queryNorm.includes(marker))) return false;
  return score < 0.95;
};

export const matchThaiFoodLocally = async (name) => {
  const portion = parsePortion(name);
  const cleanName = portion.foodName;
  const norm = normalize(cleanName);
  if (!norm) return null;

  const foods = await loadThaiFoods();
  const index = buildAliasIndex(foods);

  let best = null;
  let bestScore = 0;

  index.forEach((row) => {
    if (row.aliasNorm === norm) {
      best = { item: row.item, score: 1, matchedAlias: row.alias };
      bestScore = 1;
      return;
    }

    const score = similarity(norm, row.aliasNorm);
    if (score > bestScore) {
      bestScore = score;
      best = { item: row.item, score, matchedAlias: row.alias };
    }
  });

  if (best && bestScore >= 0.72 && !isGenericRiceMismatch(norm, best.matchedAlias, bestScore)) {
    return { ...best, portion };
  }
  return null;
};

const YUM_SALAD_EXCLUDES = ["ต้มยำ", "ข้าวยำ", "ก๋วยเตี๋ยวต้มยำ"];

const estimateYumSaladLocally = (name) => {
  const portion = parsePortion(name);
  const norm = normalize(portion.foodName);
  if (!norm) return null;
  if (YUM_SALAD_EXCLUDES.some((marker) => norm.includes(marker))) return null;
  if (!(/^ยำ/.test(norm) || norm.includes("ลาบ") || norm.includes("แซ่บ"))) return null;

  let nutrition = { calories: 280, protein: 20, carbs: 12, fat: 16 };
  if (norm.includes("กุ้ง") || norm.includes("ทะเล")) {
    nutrition = { calories: 220, protein: 18, carbs: 14, fat: 10 };
  } else if (norm.includes("หมู") || norm.includes("คอหมู")) {
    nutrition = { calories: 310, protein: 20, carbs: 10, fat: 22 };
  } else if (norm.includes("ไก่") || norm.includes("แซ่บ")) {
    nutrition = { calories: 285, protein: 22, carbs: 12, fat: 16 };
  }

  return {
    name: portion.originalName,
    ...nutrition,
    estimated: true,
    estimateSource: "nlp_parse_local",
    matchedReference: "ประมาณจากเมนูยำ/ลาบ",
  };
};

export const estimateFoodLocally = async (name) => {
  const portion = parsePortion(name);
  const thaiMatch = await matchThaiFoodLocally(name);

  if (thaiMatch) {
    const scaled = scaleNutrition(thaiMatch.item, thaiMatch.portion || portion, thaiMatch.item);
    const portionNote = formatPortionNote(thaiMatch.portion || portion);
    return {
      name: portion.originalName,
      baseName: thaiMatch.matchedAlias,
      category: thaiMatch.item.category,
      defaultServingGrams: thaiMatch.item.defaultServingGrams,
      defaultServingLabel: thaiMatch.item.defaultServingLabel,
      ...scaled,
      estimated: true,
      estimateSource: "thai_reference_local",
      matchedReference: portionNote ? `${thaiMatch.matchedAlias} · ${portionNote}` : thaiMatch.matchedAlias,
      portionNote,
    };
  }

  const nlpParsed = await parseFoodTextLocally(name, async (foodName) => matchThaiFoodLocally(foodName));
  if (nlpParsed) {
    const scaled = scaleNutrition(nlpParsed, portion);
    const portionNote = formatPortionNote(portion);
    return {
      ...nlpParsed,
      name: portion.originalName,
      baseName: nlpParsed.dishLabel || portion.foodName || portion.originalName,
      ...scaled,
      portionNote,
    };
  }

  const yumEstimate = estimateYumSaladLocally(name);
  if (yumEstimate) {
    const scaled = scaleNutrition(yumEstimate, portion);
    const portionNote = formatPortionNote(portion);
    return {
      ...yumEstimate,
      ...scaled,
      portionNote,
    };
  }

  return {
    name: portion.originalName,
    ...DEFAULT_NUTRITION,
    estimated: true,
    estimateSource: "rule_local",
  };
};

export const estimateCustomFood = async (name) => {
  const cleanName = String(name || "").trim();
  if (!cleanName) return null;

  try {
    const res = await fetch(`${getApiBaseUrl()}/api/foods/estimate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name: cleanName }),
    });
    if (res.ok) {
      const data = await res.json();
      return { ...data, custom: true };
    }
  } catch (err) {
    console.warn("Food estimate API unavailable, using local reference:", err);
  }

  return { ...(await estimateFoodLocally(cleanName)), custom: true };
};

export const isGenericFoodEstimate = (estimate) => {
  const source = estimate?.estimateSource;
  return source === "rule_local"
    || source === "rule"
    || source === "rule_default";
};

export const canSaveFoodEntry = (food) => {
  if (!food) return false;
  if (food.id != null) return true;
  const name = food.baseName || stripPortionSuffix(food.name) || food.name || "";
  if (looksLikeInvalidFoodName(name)) return false;
  if (isGenericFoodEstimate(food)) {
    return food.nutritionUnverified === true && Number(food.calories) > 0;
  }
  return isVerifiedFoodEntry(food);
};

export const canEstimateCustomFood = (estimate, name) => {
  if (!estimate) return false;
  const portion = parsePortion(name);
  const baseName = portion.foodName || name;
  const entry = { ...estimate, name, baseName };
  if (canSaveFoodEntry(entry)) return true;
  if (isGenericFoodEstimate(entry) && Number(entry.calories) > 0 && !looksLikeInvalidFoodName(baseName)) {
    return true;
  }
  return false;
};

export const prepareCustomFoodEstimate = (estimate, name) => {
  const portion = parsePortion(name);
  const baseName = portion.foodName || name;
  const entry = { ...estimate, name, baseName };
  if (isGenericFoodEstimate(entry)) {
    return {
      ...entry,
      nutritionUnverified: true,
      matchedReference: entry.matchedReference || "ประมาณคร่าวๆ — ปรับปริมาณได้ใน modal",
    };
  }
  return entry;
};

const INVALID_FOOD_NAME_CHARS = /[(){}[\]<>@#$%^&*+=|\\;:"/?!~`]/;

export const looksLikeInvalidFoodName = (name) => {
  const text = String(name || "").trim();
  if (text.length < 2) return true;
  if (INVALID_FOOD_NAME_CHARS.test(text)) return true;

  const thaiChars = (text.match(/[\u0E00-\u0E7F]/g) || []).length;
  const thaiRatio = thaiChars / text.replace(/\s/g, "").length;
  const latinOnly = /^[a-zA-Z0-9\s\-'.]+$/;

  if (thaiChars === 0 && !latinOnly.test(text)) return true;
  if (thaiChars > 0 && thaiRatio < 0.55) return true;

  const thaiOnly = text.replace(/[^\u0E00-\u0E7F]/g, "");
  if (thaiOnly.length >= 4) {
    const hasVowel = /[\u0E30-\u0E3A\u0E40-\u0E4E]/.test(thaiOnly);
    const longConsonantRun = /[\u0E01-\u0E2E]{5,}/.test(thaiOnly);
    if (!hasVowel && longConsonantRun) return true;
  }

  return false;
};

const VERIFIED_ESTIMATE_SOURCES = new Set([
  "thai_reference",
  "thai_reference_local",
  "database",
  "database_fuzzy",
  "nlp_parse",
  "nlp_parse_local",
  "calorieninjas",
  "llm_parse",
]);

export const isVerifiedFoodEntry = (food) => {
  if (!food) return false;
  if (food.id != null) return true;
  if (food.nutritionVerified === true) return true;
  const source = food.estimateSource;
  if (source && VERIFIED_ESTIMATE_SOURCES.has(source)) return true;
  if (!source && !food.custom && !food.estimated) {
    return Number(food.calories) > 0;
  }
  return false;
};

export const isUnverifiedFoodEntry = (food) => {
  if (!food) return false;
  if (food.nutritionUnverified === true) return true;
  if (isVerifiedFoodEntry(food)) return false;
  return Boolean(food.custom || food.estimated || isGenericFoodEstimate(food));
};

export const getUnverifiedMealItems = (mealList) =>
  (mealList || []).filter(isUnverifiedFoodEntry);

export const mealHasUnverifiedNutrition = (mealList) =>
  getUnverifiedMealItems(mealList).length > 0;

export const mealIsFullyUnverified = (mealList) => {
  const list = mealList || [];
  return list.length > 0 && list.every(isUnverifiedFoodEntry);
};

export const getEstimateSourceLabel = (source) => {
  if (source === "thai_reference" || source === "thai_reference_local") return "อ้างอิงเมนูไทย";
  if (source === "nlp_parse" || source === "nlp_parse_local") return "AI แกะวัตถุดิบ (NLP)";
  if (source === "llm_parse") return "AI แกะวัตถุดิบ (LLM)";
  if (source === "database" || source === "database_fuzzy") return "อ้างอิงฐานข้อมูล";
  if (source === "calorieninjas") return "อ้างอิง CalorieNinjas";
  if (source === "rule" || source === "rule_default") return "ประมาณจากชื่อเมนู";
  return "ประมาณการ";
};

File: backend/food_estimator.py
import re
from difflib import SequenceMatcher

from nutrition_api import fetch_calorieninjas
from nlp_food_parser import estimate_from_text_parsing
from portion_parser import format_portion_note, parse_portion, scale_nutrition
from thai_food_matcher import match_thai_food

DEFAULT_NUTRITION = {"calories": 450, "protein": 20, "carbs": 52, "fat": 16}

# fallback กรณีไม่ match ฐานข้อมูลเมนูไทย (เรียงจากเฉพาะเจาะจง → ทั่วไป)
FOOD_RULES = [
    (["ข้าวกะเพรา", "กะเพรา"], {"calories": 589, "protein": 28, "carbs": 75, "fat": 22}),
    (["ข้าวผัด"], {"calories": 534, "protein": 14, "carbs": 82, "fat": 16}),
    (["ข้าวมัน"], {"calories": 602, "protein": 24, "carbs": 68, "fat": 28}),
    (["ผัดไทย"], {"calories": 460, "protein": 14, "carbs": 68, "fat": 14}),
    (["ส้มตำ"], {"calories": 280, "protein": 6, "carbs": 52, "fat": 8}),
    (["ยำ", "ลาบ", "แซ่บ"], {"calories": 280, "protein": 20, "carbs": 12, "fat": 16}),
    (["ต้มยำ"], {"calories": 200, "protein": 16, "carbs": 12, "fat": 10}),
    (["แกงเขียวหวาน"], {"calories": 380, "protein": 22, "carbs": 12, "fat": 28}),
    (["ก๋วยเตี๋ยว"], {"calories": 380, "protein": 22, "carbs": 52, "fat": 10}),
    (["สลัด"], {"calories": 180, "protein": 12, "carbs": 14, "fat": 8}),
    (["ไก่ทอด"], {"calories": 540, "protein": 32, "carbs": 22, "fat": 36}),
    (["หมูทอด", "หมูกรอบ"], {"calories": 520, "protein": 28, "carbs": 8, "fat": 42}),
    (["ไข่เจียว"], {"calories": 250, "protein": 14, "carbs": 4, "fat": 20}),
    (["โจ๊ก"], {"calories": 320, "protein": 18, "carbs": 45, "fat": 8}),
    (["กล้วย", "banana"], {"calories": 105, "protein": 1.3, "carbs": 27, "fat": 0.4}),
    (["ส้ม", "orange"], {"calories": 62, "protein": 1.2, "carbs": 15.4, "fat": 0.2}),
    (["มะม่วง", "mango"], {"calories": 135, "protein": 1.1, "carbs": 35, "fat": 0.6}),
    (["มะละกอ", "papaya"], {"calories": 59, "protein": 0.9, "carbs": 15, "fat": 0.2}),
    (["แอปเปิ้ล", "apple"], {"calories": 95, "protein": 0.5, "carbs": 25, "fat": 0.3}),
    (["องุ่น", "grape"], {"calories": 104, "protein": 1.1, "carbs": 27, "fat": 0.2}),
    (["แตงโม", "watermelon"], {"calories": 86, "protein": 1.7, "carbs": 22, "fat": 0.4}),
    (["สับปะรด", "pineapple"], {"calories": 82, "protein": 0.9, "carbs": 22, "fat": 0.2}),
    (["ทุเรียน", "durian"], {"calories": 147, "protein": 1.5, "carbs": 27, "fat": 5}),
    (["มังคุด", "mangosteen"], {"calories": 63, "protein": 0.6, "carbs": 16, "fat": 0.2}),
    (["ลำไย", "longan"], {"calories": 120, "protein": 1.3, "carbs": 31, "fat": 0.1}),
    (["ลิ้นจี่", "lychee"], {"calories": 125, "protein": 1.6, "carbs": 31, "fat": 0.4}),
    (["ฝรั่ง", "guava"], {"calories": 37, "protein": 1.4, "carbs": 8, "fat": 0.5}),
]


def _normalize(text):
    return re.sub(r"\s+", " ", str(text or "").strip().lower())


def _nutrition_from_row(row):
    return {
        "calories": round(float(row.get("calories") or 0)),
        "protein": round(float(row.get("protein") or 0), 1),
        "carbs": round(float(row.get("carbs") or row.get("carb") or 0), 1),
        "fat": round(float(row.get("fat") or 0), 1),
    }


def _best_db_match(name, cursor):
    clean_name = re.sub(r"\s+", " ", str(name or "").strip())
    norm = _normalize(clean_name)
    if not norm or cursor is None:
        return None

    candidates = []
    try:
        cursor.execute("SELECT * FROM foods WHERE name LIKE %s LIMIT 20", (f"%{clean_name}%",))
        candidates.extend(cursor.fetchall())

        for token in [t for t in re.split(r"[\s,/]+", clean_name) if len(t) >= 2][:4]:
            cursor.execute("SELECT * FROM foods WHERE name LIKE %s LIMIT 8", (f"%{token}%",))
            candidates.extend(cursor.fetchall())
    except Exception:
        return None

    if not candidates:
        return None

    unique = {}
    for row in candidates:
        key = row.get("id") or row.get("name")
        unique[key] = row

    best_row = None
    best_score = 0.0
    for row in unique.values():
        row_name = str(row.get("name") or "")
        row_norm = _normalize(row_name)
        if row_norm == norm:
            return row, 1.0, row_name

        score = SequenceMatcher(None, norm, row_norm).ratio()
        if norm in row_norm or row_norm in norm:
            overlap = min(len(norm), len(row_norm)) / max(len(norm), len(row_norm))
            score = max(score, 0.84 + overlap * 0.16)

        if score > best_score:
            best_score = score
            best_row = row

    if best_row and best_score >= 0.68:
        return best_row, best_score, str(best_row.get("name") or "")
    return None


def _keyword_estimate(name):
    lowered = name.lower()
    for keywords, nutrition in FOOD_RULES:
        if any(keyword.lower() in lowered or keyword in name for keyword in keywords):
            return dict(nutrition), "rule"
    return dict(DEFAULT_NUTRITION), "rule_default"


def _build_result(name, nutrition, source, matched_alias=None, portion_note=None):
    payload = {
        "name": name,
        "calories": round(nutrition["calories"]),
        "protein": round(float(nutrition["protein"]), 1),
        "carbs": round(float(nutrition["carbs"]), 1),
        "fat": round(float(nutrition["fat"]), 1),
        "estimated": True,
        "estimateSource": source,
    }
    if matched_alias:
        payload["matchedReference"] = matched_alias
    if portion_note:
        payload["portionNote"] = portion_note
    return payload


def estimate_food_from_name(name, cursor=None):
    clean_name = re.sub(r"\s+", " ", str(name or "").strip())
    if not clean_name:
        return None

    portion = parse_portion(clean_name)
    match_name = portion["foodName"]
    portion_note = format_portion_note(portion)

    thai_match = match_thai_food(match_name)
    if thai_match and thai_match["score"] >= 0.72:
        item = thai_match["item"]
        scaled = scale_nutrition(item, portion, item)
        reference = thai_match["matchedAlias"]
        if portion_note:
            reference = f"{reference} · {portion_note}"
        return _build_result(
            clean_name,
            scaled,
            "thai_reference",
            reference,
            portion_note,
        )

    db_match = _best_db_match(match_name, cursor)
    if db_match:
        row, score, matched_name = db_match
        scaled = scale_nutrition(_nutrition_from_row(row), portion)
        reference = matched_name
        if portion_note:
            reference = f"{matched_name} · {portion_note}"
        return _build_result(
            clean_name,
            scaled,
            "database" if score >= 0.9 else "database_fuzzy",
            reference,
            portion_note,
        )

    nlp_result = estimate_from_text_parsing(match_name)
    if nlp_result:
        scaled = scale_nutrition(nlp_result, portion)
        payload = _build_result(
            clean_name,
            scaled,
            nlp_result.get("estimateSource", "nlp_parse"),
            nlp_result.get("matchedReference"),
            portion_note,
        )
        if nlp_result.get("parsedIngredients"):
            payload["parsedIngredients"] = nlp_result["parsedIngredients"]
            payload["parseMethod"] = nlp_result.get("parseMethod")
            payload["dishLabel"] = nlp_result.get("dishLabel")
        return payload

    api_result = fetch_calorieninjas(clean_name)
    if api_result:
        scaled = scale_nutrition(api_result, portion)
        reference = api_result.get("matchedQuery")
        if portion_note and reference:
            reference = f"{reference} · {portion_note}"
        return _build_result(
            clean_name,
            scaled,
            api_result.get("estimateSource", "calorieninjas"),
            reference,
            portion_note,
        )

    nutrition, source = _keyword_estimate(match_name)
    scaled = scale_nutrition(nutrition, portion)
    return _build_result(clean_name, scaled, source, portion_note=portion_note)
